packages/interactivity/src/store.ts

```diff
@@ -50,6 +50,62 @@ interface StoreOptions {
 	lock?: boolean | string;
 }
 
+type Prettify< T > = { [ K in keyof T ]: T[ K ] } & {};
+type DeepPartial< T > = T extends object
+	? { [ P in keyof T ]?: DeepPartial< T[ P ] > }
+	: T;
+type DeepPartialState< T extends { state: object } > = Omit< T, 'state' > & {
+	state?: DeepPartial< T[ 'state' ] >;
+};
+type ConvertGeneratorToPromise< T > = T extends (
+	...args: infer A
+) => Generator< any, infer R, any >
+	? ( ...args: A ) => Promise< R >
+	: never;
+type ConvertGeneratorsToPromises< T > = {
+	[ K in keyof T ]: T[ K ] extends ( ...args: any[] ) => any
+		? ConvertGeneratorToPromise< T[ K ] > extends never
+			? T[ K ]
+			: ConvertGeneratorToPromise< T[ K ] >
+		: T[ K ] extends object
+		? Prettify< ConvertGeneratorsToPromises< T[ K ] > >
+		: T[ K ];
+};
+type ConvertPromiseToGenerator< T > = T extends (
+	...args: infer A
+) => Promise< infer R >
+	? ( ...args: A ) => Generator< any, R, any >
+	: never;
+type ConvertPromisesToGenerators< T > = {
+	[ K in keyof T ]: T[ K ] extends ( ...args: any[] ) => any
+		? ConvertPromiseToGenerator< T[ K ] > extends never
+			? T[ K ]
+			: ConvertPromiseToGenerator< T[ K ] >
+		: T[ K ] extends object
+		? Prettify< ConvertPromisesToGenerators< T[ K ] > >
+		: T[ K ];
+};
+
+/**
+ * Creates a generator that yields a promise and returns its resolved value.
+ *
+ * This utility function is used to convert promises into generators, which
+ * can be useful when working with asynchronous actions.
+ *
+ * @since 6.7.0
+ *
+ * @param promise The promise to be converted into a generator.
+ *
+ * @return A generator that yields the promise and returns its resolved value.
+ */
+export function typed< T >(
+	promise: Promise< T >
+): Generator< Promise< T >, T, T > {
+	return ( function* () {
+		return yield promise;
+	} )();
+}
```

Comment:
I have some concerns about this.

(minor) The name doesn't indicate anything meaningful, it needs improvement.

(minor) The changes in how this is supposed to be used needs an example in the JSDoc.

More importantly, this function introduces a pattern that is now exclusive to TypeScript. I believe it exists almost entirely to improve types, but its implementation certainly does things. I need to think about alternatives. Related: https://github.com/microsoft/TypeScript/issues/57671, https://github.com/microsoft/TypeScript/issues/43632.

Now we have two patterns to do exactly the same thing but are subtly different (one needs is wrapped in the function and needs another `*`):

```ts
store( 'foo', {
  x: function* () {
    let val;
    try {
      val = yield Promise.reject(1);
    } catch (e) {
      console.error(e);
    }
    console.log(val);
  },
  y: function* () {
    let val;
    try {
      val = yield* I.typed(/** @type {Promise<number>} */ (Promise.reject(2)));
    } catch (e) {
      console.error(e);
    }
    console.log(val);
  },
} );
```

If we go with this approach, it seems like we should discourage the plain `yield Promise.resolve(1)` and rely on this function (with an improved name).

Reply:
That said, I also don't think there's a rush to ship this utility if we're not sure about it 🙂

Reply:
@juanmaguitar and @sirreal, one quick question. If we move `typed()` with accompanying unit tests to a follow up PR, is everything else ready to go? In the documentation, the examples that use `typed()` would also need to get removed but they contain a note that it requires WP 6.7 so they aren't too much useful for now anyway. If we would land the rest today, then the new guide would be available in the dev docs at the same time when all the changes to packages would get published to npm (will happen overnight).

Reply:
> If we move typed() with accompanying unit tests to a follow up PR, is everything else ready to go?

In my opinion that's the best plan 👍

Reply:
I'll split out a separate PR today.

Reply:
https://github.com/WordPress/gutenberg/pull/65439 matches this branch right now, but will be used to keep the work on the `typed` function.

Reply:
@sirreal @gziolo now that https://github.com/WordPress/gutenberg/pull/65439 replicates this PR 
- Shouldn't https://github.com/WordPress/gutenberg/pull/65439 contains only changes under `packages/interactivity` 
- Shouldn't this PR contains only changes under `docs` & `packages/create-block-interactive-template` ?

Reply:
https://github.com/WordPress/gutenberg/pull/65439 isn't ready, changes from this branch haven't been removed from that branch. That will be done in the future, it isn't a priority for now.

There are changes and improvements to TypeScript support in `packages/interactivity` that are ready to land in this PR, so it's expected that there are changes to that directory in this branch.

Reply:
Oh, I see! Thanks for the clarification.
I think it's good to go, then.